### ISSUE TITLE
persist: design for efficient access to historical data

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#26938]
Tracks shaking out a design for https://github.com/MaterializeInc/materialize/issues/17492, but _not_ the actual implementation work.

**Comment by danhhz** [comment]
There are two potential designs here with pretty different tradeoffs. To decide between them, we'd likely need to prototype them both.

Idea: Scrub through historical SeqNos
- Because we offload historical state diffs into s3, it is currently very cheap to have a very long seqno hold. (We still today see "leaked" seqno holds in prod with no visible customer effect.)
- So, be aggressive about downgrading the since and unblocking compaction for the most recently written data, but build the ability to translate a historical since into a historical seqno hold to be able to recover historical times.
- Then, maintain an index in the most recent version of state of frontiers -> seqnos at which they can be read at original fidelity. Use this to efficiently start at an appropriate, historical version of persist state and step it forward seqno by seqno to advance up to realtime.
- Hesitation: persist "one state" has been a wild success at preventing the memory blowups we had in early persist days. This would be intentionally re-introducing multiple copies of the same shard's state.
- Hesitation: everything in persist right now can work equally well at the latest version of state, so it's always safe to advance forward when you find out about a new seqno. This design would change that, and so might shake out some places where we've baked that assumption into code.

Idea: Keep compaction history around in SpineBatch
- NB: Ben has more detailed thoughts here than I do, so I may get some of this slightly wrong!
- SpineBatch already is recursive and keeps track of outstanding compactions. When one of them finishes we swap in the output and remove the inputs.
- We could instead leave the inputs around. This would by definition give us the ability to read back data at any time we could in the past.
- This has the benefit of always working on the latest version of state, like the rest of persist.
- Hesitation: Would make state bigger. We already have this issue for large shards, stats, inline writes and need to solve it, so maybe cross this one off.
- Hesitation: Potentially makes Spine more complicated? Spine is already complicated enough, so we'd be pretty sensitive to adding more to it. Seems like leaving more stuff around in SpineBatch is relatively orthogonal to its complexity, but this is one of the things a prototype would shake out for us.

The original motivation for pulling out "design" into a separate issue was so we could track it as part of Q2 work. We haven't produced a design doc or prototype, but now that it's Q3, no real value no in keeping this around separate from [he overall issue](https://github.com/MaterializeInc/materialize/issues/17492), so I'll go ahead and close it.
